django/db/models/functions/text.py

```diff
@@ -65,83 +65,49 @@ def as_sqlite(self, compiler, connection, **extra_context):
         return super().as_sql(compiler, connection, function="CHAR", **extra_context)
 
 
-class ConcatPair(Func):
+class Concat(Func):
     """
-    Concatenate two arguments together. This is used by `Concat` because not
-    all backend databases support more than two arguments.
+    Concatenate expressions together. Expressions that are nullable
+    are coalesced to an empty string and non-textual expressions are
+    casted to strings.
     """
 
-    function = "CONCAT"
+    function = None
+    template = "%(expressions)s"
+    arg_joiner = " || "
 
-    def as_sqlite(self, compiler, connection, **extra_context):
-        coalesced = self.coalesce()
-        return super(ConcatPair, coalesced).as_sql(
-            compiler,
-            connection,
-            template="%(expressions)s",
-            arg_joiner=" || ",
-            **extra_context,
+    def __init__(self, *expressions, **extra):
+        if len(expressions) < 2:
+            raise ValueError("Concat must take at least two expressions")
+        super().__init__(*expressions, **extra)
+
+    def as_mysql(self, compiler, connection):
+        # MySQL || operator is an alias for OR unless PIPES_AS_CONCAT is enabled.
+        return Func(Value(""), *self.source_expressions, function="CONCAT_WS").as_sql(
+            compiler, connection
         )
 
-    def as_postgresql(self, compiler, connection, **extra_context):
+    def as_sql(self, compiler, connection, **extra_context):
         copy = self.copy()
         copy.set_source_expressions(
             [
-                Cast(expression, TextField())
+                Coalesce(
```

Comment:
I figured that the `Coalesce` is not necessary on backends that treat nulls as empty strings (Oracle) but adding a check for that would have made the logic even more complex.